### PR TITLE
Allow both terminal and running costs

### PR DESCRIPTION
**Pull request by cfinlay**
Summary of edits:

1) In certain applications there are both running and terminal costs. This change modifies the adjoint dynamics so that both can be used.

2) I've also changed the running cost so that it may depend on the dynamic's function (dx/ds).

3) Finally I modified the context manager for gradient computations, so that I could define dynamics that are a gradient of a scalar potential.

These last two are not so important, but are nice to have. Take them or leave them. However IMHO the first edit is quite important.

### CODE REVIEW

**Comment by Zymrael** [comment]
Thank you @cfinlay! I'll take a look and integrate the changes with the development branch `new_release_dev`. 

We actually also carried out a small redesign of `NeuralDE`'s context manager and API in preparation for the new release (branch `new_release_dev`), and in doing so implemented [1].

I agree that [2] is nice to have, so it will definitely be included

[3] is related to several design choices affecting models like [Stable Neural ODEs](https://arxiv.org/pdf/2003.08063.pdf), [Hamiltonian Neural ODEs](https://arxiv.org/pdf/1906.01563.pdf), [Lagrangian Neural ODEs](https://arxiv.org/pdf/2003.04630.pdf). Currently, these models are designed as wrappers to `nn.Modules` to act as a interface between `DEFunc` and the generic `nn.Module`. This is also the level of abstraction where, ultimately, `continuous normalizing flows` (CNFs) are planned to be included. 

I'll open a new issue on [3] to encourage more discussion and receive additional feedback as I expect CNFs to be a popular feature within the `torchdyn` ecosystem.

**Comment by cfinlay** [comment]
Great, sounds good! Thanks for taking a look. I'm glad you've already implemented [1].

[3] is very interesting. It would be great if the vector field could be defined as $dx/ds = \nabla \phi(x)$, where $\phi$ is a scalar function. The issue that I was running into is that the autograd context manager wasn't allowing gradient computations in the dynamics. So if you fix that, that would be great. This pull request is one way of fixing it, but I'm sure there are many other ways to go about it too.
